Added real and imag views as tensor attributes

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39099 Add view_as_real for complex tensors
* #39065 Remove copy_imag and copy_real methods
* **#39033 Added real and imag views as tensor attributes**

Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

fixes https://github.com/pytorch/pytorch/issues/35494
Differential Revision: [D21767542](https://our.internmc.facebook.com/intern/diff/D21767542)

anjali411 committed:
Added real and imag views as tensor attributes

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

albanD commented on torch/_tensor_docs.py:
```diff
@@ -3733,6 +3733,36 @@ def callable(a, b) -> number
+add_docstr_all('real',
+               r"""
+Returns a new tensor containing real values of the :attr:`self` tensor.
+The returned tensor shares the same data as :attr:`self`. For complex tensors,
```
nit: can you use "memory" instead of "data" here to avoid any confusion with `.data`.
Same below.

albanD commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -135,15 +136,32 @@ Tensor angle(const Tensor& self) {
+    if (self.numel() == 0) {
+      return at::empty({0}, self.options().dtype(c10::toValueType(self.scalar_type())));
+    } else {
+      auto float_tensor = at::native::view_complex_as_float(self);
```
Looking at the implementation of `view_complex_as_float`, I don't think you want to use it here.
The returned Tensor does not own its memory. So as soon as the complex Tensor goes out of scope, the view will become invalid.
Since you return this view to the user, it should be owning it's memory or have a very very big warning in the doc:
```
def eval_my_fn(inp):
    complex_out = my_fn(inp)
    return complex_out.real #Only return the real part

out = eval_my_fn(inp)
print(out) # This will segfault.

anjali411 replied:
I think it's a good point and I will update the code to make the tensor from the same storage as self, but the case you mentioned above doesn't segfault with this code:
```
>>> def eval_my_fn(inp):
...     complex_out = inp.conj()
...     return complex_out.real
...
>>> out = eval_my_fn(torch.randn(2,2, dtype=torch.cfloat))
>>> print(out)
tensor([[-9.1000e-17,  5.6052e-45],
        [ 1.4013e-45,  0.0000e+00]])
```

albanD replied:
Most likely because you're lucky and the memory is not returned to the OS yet. But I'm sure that if you play around with sizes or reallocate a new Tensor with the same size you can get very bad stuff to happen.

albanD commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -135,15 +136,32 @@ Tensor angle(const Tensor& self) {
+      return at::empty({0}, self.options().dtype(c10::toValueType(self.scalar_type())));
+    } else {
+      auto float_tensor = at::native::view_complex_as_float(self);
+      return at::narrow(float_tensor, float_tensor.dim() - 1, 0, 1).view(self.sizes());
```
You should use `at::select()` here, no need to have an extra argument and a dimension of size 1.

albanD commented on test/test_torch.py:
```diff
@@ -17539,45 +17539,33 @@ def is_view_of(self, base, other):
+    @dtypes(torch.float, torch.double, torch.long)
+    def test_real_self(self, device, dtype):
+        t = torch.ones((5, 5), dtype=dtype, device=device)
+        s = t.real
```
`torch.real()` should still work right? I think you still want to test it.

albanD commented on test/test_torch.py:
```diff
@@ -17539,45 +17539,33 @@ def is_view_of(self, base, other):
-    def test_real_self(self, device):
-        t = torch.ones((5, 5), device=device)
-        s = torch.real(t)
+    @dtypes(torch.float, torch.double, torch.long)
```
Why only test these dtypes and not all of them?

albanD commented on test/test_torch.py:
```diff
@@ -17539,45 +17539,33 @@ def is_view_of(self, base, other):
-            v[0] = 0
-            self.assertEqual(t.float()[0], v[0])
-            self.assertTrue(t[0] == complex(0, 1))
+    @dtypes(torch.cfloat, torch.cdouble)
```
I think you want to keep the check that it raises the expected error for all the other dtypes.

albanD reviewed (changes requested):
My main concern is with `view_complex_as_float`. I don't think you should use it in this context.

mruberry commented on test/test_torch.py:
```diff
@@ -17539,45 +17539,33 @@ def is_view_of(self, base, other):
+        re = t.real
+        exp = torch.from_numpy(t.numpy().real)
+        self.assertEqual(re, exp)
+        assert(t.storage().data_ptr() == re.storage().data_ptr())
```
`self.assertEqual(t.storage().data_ptr(), re.storage().data_ptr())`

mruberry replied:
Also: can you not use is_view_of here?

anjali411 replied:
`is_view_of` also check the autograd side of things (`other._base is not base`), so will update this test to use is_view_of when the autograd code is modified to handle `.real` and `.imag`

mruberry replied:
OK. If .real and .imag don't work well with autograd, should they be prevented on tensors requiring a gradient? Much like how we don't allow .numpy() on those tensors, either (you have to do tensor.detach().cpu().numpy(), for example).

albanD replied:
The plan is to fix that behavior in another PR in this stack.
I don't think we need to add this check to remove it in another PR in the same stack that will be merged at the same time anyway.

mruberry commented on test/test_torch.py:
```diff
@@ -17539,45 +17539,33 @@ def is_view_of(self, base, other):
+        im = t.imag
+        exp = torch.from_numpy(t.numpy().imag)
+        self.assertEqual(im, exp)
+        assert(t.storage().data_ptr() == im.storage().data_ptr())
```
same comment as above

mruberry commented on test/test_torch.py:
```diff
@@ -17539,45 +17539,33 @@ def is_view_of(self, base, other):
+    def test_real_imag_view(self, device, dtype):
+        t = torch.randn(3, 3, dtype=dtype, device=device)
+        re = t.real
+        exp = torch.from_numpy(t.numpy().real)
```
Comparing it to NumPy here is a great idea.

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -3733,6 +3733,36 @@ def callable(a, b) -> number
 
+add_docstr_all('real',
+               r"""
+Returns a new tensor containing real values of the :attr:`self` tensor.
```
Not a new tensor for complex values, either.

anjali411 replied:
for complex tensors, `tensor.real` returns a new tensor with the same storage as the tensor it's being called on

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -3733,6 +3733,36 @@ def callable(a, b) -> number
+               r"""
+Returns a new tensor containing imaginary values of the :attr:`self` tensor.
+The returned tensor shares the same data as :attr:`self`. For complex tensors,
+the returned tensor has the corresponding real dtype.
```
Should probably make clearer that this only works on complex tensors. You could add a warning that it's not implemented for non-complex tensors, or change the language to be something like:

"Returns a view of the complex tensor's imaginary values..."

mruberry commented:
jit failures are, unfortunately, real. Otherwise this looks great to me. Meta question: what happens if I take the real view of a complex tensor that requires grad, then perform some computations with it?

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

albanD commented:
@mruberry the autograd behavior will be what you expect at the end of the stack: the same as view between real Tensors.
To make reviewing easier, we decided to split the changes in multiple PRs in a single stack.

albanD commented on aten/src/ATen/native/ComplexHelper.h:
```diff
@@ -18,23 +18,15 @@ inline std::vector<int64_t> computeStrideForComplex(IntArrayRef oldstride) {
-      double* data = reinterpret_cast<double*>(self.data_ptr<std::complex<double>>());
-      return at::from_blob(data, new_sizes, new_strides, self.options().dtype(at::kDouble));
-    }
+    return at::empty({0}, self.options().dtype(float_type)).set_(self.storage(), 0, new_sizes, new_strides);
```
I think you want to use the storage offset from self here and not `0`. The offset is already included when you take data_ptr but is not when you take the storage.

anjali411 replied:
good catch! updated the input storage offset and added a test as well

albanD commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -135,15 +136,33 @@ Tensor angle(const Tensor& self) {
+  } else {
+    TORCH_CHECK(false, "real is not implemented for tensors with non-complex dtypes.");
+    // Note: unreachable
+    return self;
```
I know this is not reachable but please make it consistent between real and imag.

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -516,6 +516,24 @@ static PyObject * THPVariable_device(THPVariable* self, void *unused) {
+{
+  HANDLE_TH_ERRORS
+  auto& self_ = self->cdata;
+  auto real = at::native::real(self_);
```
nit: when we add autograd this should become "at::real()" no?

albanD reviewed (changes requested):
One missing storage offset and then it's good to go for me.

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

Differential Revision: [D21767542](https://our.internmc.facebook.com/intern/diff/D21767542)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -135,15 +136,29 @@ Tensor angle(const Tensor& self) {
-  return at::zeros_like(self);
+  if (self.is_complex()) {
+    auto float_tensor = at::native::view_complex_as_float(self);
+    if (self.numel() == 0) {
```
This special case looks strange!

anjali411 replied:
yeah good catch! removed the special case since it was unnecessary and also inconsistent with the definition where we state that `view_complex_as_float` returns a tensor of size `m1 x m2 x ... mn x 2` for a tensor of size `m1 x m2 x ... mn` . It should remain the case for tensors with zero elements as well.

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

Differential Revision: [D21767542](https://our.internmc.facebook.com/intern/diff/D21767542)

[ghstack-poisoned]

anjali411 committed:
Update on "Added real and imag views as tensor attributes"


Added `real` and `imag` views as tensor attributes. Right now, tensor.imag is disabled for real tensors. This is because if we return a new tensor of zeros, the user would be able to update the tensor returned by tensor.imag which should not be allowed as numpy returns a read-only array, and pytorch doesn't support read-only tensors yet.

TODO in follow-up PRs:
1. add a setter for `real` and `imag`
2. add special case in codegen for `real` and `imag` backward functions. 
3. remove `copy_real` and `copy_imag` methods.

Differential Revision: [D21767542](https://our.internmc.facebook.com/intern/diff/D21767542)

[ghstack-poisoned]
